sct-commons/src/main/java/org/lfenergy/compas/sct/commons/util/ControlBlockNetworkSettingsCsvHelper.java

```diff
@@ -61,48 +57,62 @@ private Map<Criteria, Settings> readCsvFile(Reader csvSource) {
     }
 
     @Override
-    public Settings getNetworkSettings(ControlBlockAdapter controlBlockAdapter) {
+    public SettingsOrError getNetworkSettings(ControlBlockAdapter controlBlockAdapter) {
         ControlBlockEnum controlBlockEnum = controlBlockAdapter.getControlBlockEnum();
         IEDAdapter iedAdapter = controlBlockAdapter.getParentIedAdapter();
-        String systemVersion = iedAdapter.getCompasSystemVersion()
-                .map(version -> version.getMainSystemVersion() + "." + version.getMinorSystemVersion())
-                .orElse(null);
-        String systemVersionWithoutV = removeVFromSystemVersion(systemVersion);
-        TCompasIEDType iedType = iedAdapter.getCompasICDHeader().map(TCompasICDHeader::getIEDType).orElse(null);
-        TCompasIEDRedundancy iedRedundancy = iedAdapter.getCompasICDHeader().map(TCompasICDHeader::getIEDredundancy).orElse(null);
+        Optional<TCompasSystemVersion> compasSystemVersion = iedAdapter.getCompasSystemVersion();
+        if (compasSystemVersion.isEmpty()) {
+            return new SettingsOrError(null, "No private COMPAS-SystemVersion found in this IED");
+        }
+        String systemVersionWithoutV = removeVFromSystemVersion(compasSystemVersion.get());
+        Optional<TCompasICDHeader> compasICDHeader = iedAdapter.getCompasICDHeader();
+        if (compasICDHeader.isEmpty()) {
+            return new SettingsOrError(null, "No private COMPAS-ICDHeader found in this IED");
+        }
+        TCompasIEDType iedType = compasICDHeader.get().getIEDType();
+        TCompasIEDRedundancy iedRedundancy = compasICDHeader.get().getIEDredundancy();
+        BigInteger iedSystemVersionInstance = compasICDHeader.get().getIEDSystemVersioninstance();
         boolean isBayInternal = controlBlockAdapter.getName().endsWith("I");
-        return findSettings(new Criteria(controlBlockEnum, systemVersionWithoutV, iedType, iedRedundancy, isBayInternal));
+
+        Criteria criteria = new Criteria(controlBlockEnum, systemVersionWithoutV, iedType, iedRedundancy, iedSystemVersionInstance, isBayInternal);
```

Comment:
- The main reason is that HashMap search is way faster than ArrayList search. More you have lines, more the difference is significant. https://www.baeldung.com/java-hashset-arraylist-contains-performance (HashSet is implemented using HashMap keys, so the benchmarks of HashSet.contains is the same as HashMap.get)
- Another reason (more subjective) : I prefer storing the CSV file using int, byte, boolean, enum fields instead of Strings only. I am not sure it consumes less memory, but It looks less "[Stringly typed](https://www.techopedia.com/definition/31876/stringly-typed)").